Make unauthenticated fetches for WebID

NSeydoux opened the pull request:
The WebID is not necessarily the IRI of a Solid Resource, and as such it isn't correct to make an authenticated request to it. `getPodUrlAll` and `getProfileAll` now both use an explicitly unauthenticated fetch to initially get the WebID profile, and only then either use the provided authenticated fetch or use our default session heuristics.

- [X] I've added a unit test to test for potential regressions of this bug.
- [x] The changelog has been updated, if applicable.
- [X] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).

NSeydoux committed:
Make unauthenticated fetches for WebID

The WebID is not necessarily the IRI of a Solid Resource, and as such it isn't correct to make an authenticated request to it. `getPodUrlAll` and `getProfileAll` now both use an explicitly unauthenticated fetch to initially get the WebID profile, and only then either use the provided authenticated fetch or use our default session heuristics.

ThisIsMissEm reviewed (commented):
I'm gonna need to review this more on Monday, but the tests look maybe not correct at a first pass, but I could be misreading (looking on mobile)

matthieubosquet commented on src/profile/webid.ts:
```diff
@@ -91,21 +93,20 @@ export async function getProfileAll<
-  } = options;
-
+  const authFetch = options?.fetch ?? defaultFetch;
+  const webIdProfile =
+    options?.webIdProfile ??
+    ((await getSolidDataset(webId, { fetch: unauthenticatedFetch })) as T);
   const altProfileAll = (
     await Promise.allSettled(
       getAltProfileUrlAllFrom(webId, webIdProfile).map((uniqueProfileIri) =>
-        getSolidDataset(uniqueProfileIri, { fetch })
+        getSolidDataset(uniqueProfileIri, { fetch: authFetch })
```
This looks good.

matthieubosquet commented on src/profile/webid.test.ts:
```diff
@@ -459,24 +565,30 @@ describe("getPodUrlAll", () => {
+      })
+    );
+
+    const mockedFetch = jest.fn(fetch).mockResolvedValueOnce(
+      new Response(await triplesToTurtle(toRdfJsQuads(altProfileAll[0])), {
+        headers: {
+          "Content-Type": "text/turtle",
+        },
+        url: "https://some.alt-profile",
+      } as ResponseInit)
+    );
```
Maybe in the expect section, you could add something like:

```js
expect(mockedFetch).toHaveBeenCalledTimes(1);
expect(mockedUnauthFetch).toHaveBeenCalledTimes(1);
```

Just to be more explicit about testing what was called?

matthieubosquet reviewed (approved):
LGTM.

Small suggestion about the test.

ThisIsMissEm commented on src/profile/webid.test.ts:
```diff
@@ -146,22 +157,62 @@ describe("getProfileAll", () => {
+
     await getProfileAll(MOCK_WEBID);
+    // The embedded fetch should have been used.
     expect(mockedFetcher.fetch).toHaveBeenCalled();
```
I think here we actually need to test that we're using an unauthenticated fetcher for the first call, and then an authenticated fetcher for the rest, and actually assert that, or at least check that `mockedFetcher.fetch` has been called N number of times (2 if I'm reading this right)

NSeydoux replied:
Should be fixed in fe6c5ea8f2b4419726d34b731398ab13e40a91cf

ThisIsMissEm commented on src/profile/webid.test.ts:
```diff
@@ -146,22 +157,62 @@ describe("getProfileAll", () => {
     await getProfileAll(MOCK_WEBID, { fetch: mockedFetch });
-    expect(mockedFetch).toHaveBeenCalled();
+    expect(mockedFetch).not.toHaveBeenCalled();
+    expect(mockedFetcher.fetch).toHaveBeenCalled();
```
Ditto here, `mockedFetch` should've been called once, the built-in fetch should've been called once too, and we can assert the URLs that are being fetched align with our expectations.

ThisIsMissEm commented on src/profile/webid.ts:
```diff
@@ -91,21 +93,20 @@ export async function getProfileAll<
+  const authFetch = options?.fetch ?? defaultFetch;
+  const webIdProfile =
+    options?.webIdProfile ??
+    ((await getSolidDataset(webId, { fetch: unauthenticatedFetch })) as T);
```
Aside, it might be nice to change `getSolidDataset` to accept `T` as a generic, such that we could just do:
```suggestion
    (await getSolidDataset<T>(webId, { fetch: unauthenticatedFetch }));
```

NSeydoux replied:
Actually, I'd argue that this is an issue with the function signature rather than `getSolidDataset`: in the case when `getSolidDataset` is needed to fetch the WebID, the generic type is unbound, and the return type should be `SolidDataset & WithServerResourceInfo`, which is the return type of `getSolidDataset` in the first place. I'll see if this can be cleaned up with a little overload.

NSeydoux replied:
That was easy enough: https://github.com/inrupt/solid-client-js/pull/1530

ThisIsMissEm reviewed (commented):
The tests do read a bit strange to me, and aren't particularly clear, I think we can definitely add more clarity to them, whether now or in future work.

NSeydoux committed:
Add more assertions to the tests

Check the number of times each fetch (auth and unauth) have been called, and for which resources.

NSeydoux committed:
Updated the changelog

NSeydoux committed:
Fix linting issues

ThisIsMissEm commented on src/profile/webid.test.ts:
```diff
@@ -371,28 +467,118 @@ describe("getPodUrlAll", () => {
+    );
+
+    // The WebID is explicitly fetched using the unauthenticated fetch.
+    const mockedUnauthFetcher = jest.requireMock("cross-fetch") as {
```
Can these be setup in a beforeEach?

ThisIsMissEm reviewed (approved):
Looking good, left a few small comments of things you might want to change, but otherwise we're good
